docs/packages/xstate-react/index.md

```diff
@@ -252,6 +253,59 @@ const App = ({ service }) => {
 };
 ```
 
+### `useSuspensfulGetter(state, condition, getter)`
+
+- `state` - the `state` from the first index of the return value of `useMachine()`.
+  - For example: `const [state] = useMachine(machine);`.
+- `condition` - a function that takes in `state` and returns a `boolean` that determines whether or not the resource will suspend or retrieve the value you wish to get with your `getter`.
+- `getter` - a function that takes in `state` and returns any value you want it to.
+
+#### Return value
+
+`useSuspensfulGetter` will return an object with one method — `read()`. Based on the return value of your `condition`, this method will either throw a promise or return the value from your `getter`.
+
+- If your `condition` returns a **falsey value**, `read()` will _suspend_ by throwing a promise.
+- If your `condition` returns a **truthy value**, `read()` will _return_ the result of your `getter`.
+
+In other words, once your `condition` is satisified, `read()` will stop suspending and simply give you your value that you asked for with `getter`.
+
+#### Example
+
+In this example, `Outer` is the resource creator and `Inner` is the resource consumer. In `Outer`, we use our machine `userMachine` to determine if the user is logged in and fetch their information. In most cases, getting a user's data is asynchronous and you don't often know how long it will take. Therefore, we will use `useSuspensefulGetter` to create a `resource` to suspend the section of the application that needs to know something about the user until we have access to the user — represented in code, this condition is `user !== undefined`.
+
+`useSuspensefulGetter` does not do the suspending all on its own. After the creation of this `resource`, **we must pass it down at least one level before reading it to avoid an unending loop**. Now that we've passed it down to `Inner`, we can call `resource.read()`. This will either give us access to `state.context.user` or it will suspend all elements up to the nearest `<Suspense>` element and display its `fallback` attribute instead.
```

Comment:
This is the main worrying factor for me - the limitation is real and seems like a sign that maybe there are other ways to do this.

Reply:
I see where you're coming from about this limitation. I too wonder if there is a better way to do this but according to [the examples in React's documentation on "Suspense for Data Fetching"](https://reactjs.org/docs/concurrent-mode-suspense.html#what-is-suspense-exactly), they always create the resource outside of the component where the resource is used. In our case, we need to create the resource dynamically so we have to do this within another React component.

You can also see [more examples of how this pattern works in the React documentation](https://reactjs.org/docs/concurrent-mode-patterns.html) when they demonstrate using `useState` to create the `resource` and `setResource` variables. They then follow this by passing the `resource` down at least one level because if they were to call `resource.read()` where `useState` was called, it would cause the infinite rerendering issue — a.k.a. throwing a promise causes "suspending" to occur but also blows away the state holding the resource and then rerenders. I believe that this is just a fundamental part of how Suspense and "suspending" was designed.

With all of that said, I completely understand waiting until this feature is officially no longer listed as experimental and the ideas behind Concurrent Mode and Suspense are declared solidified and ready.

Reply:
Although to be clear, while this feature, Suspense and the ability to suspend by throwing a promise, is "experimental" it has been in stable releases for quite some time now from what I can tell. It seems to be that the Concurrent Mode features as a whole are considered experimental though.

Reply:
Any energy to take a new look at this PR? Now while concurrent-mode/suspense are in alpha and announced for react 18, it may be a good point in time! :)